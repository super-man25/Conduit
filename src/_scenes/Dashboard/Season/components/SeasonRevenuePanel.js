// @flow
import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { createStructuredSelector } from 'reselect';
import { actions as seasonStatActions } from '_state/seasonStat';
import { differenceInCalendarDays } from 'date-fns';
import styled from 'styled-components';
import RevenueBreakdown from './RevenueBreakdown';
import {
  Spacing,
  Panel,
  PanelHeader,
  PanelContent,
  Flex,
  FlexItem,
  Tabbable,
  Loader,
  ChipButton,
  ChipButtonGroup,
  DateRangeFilter,
  PeriodicInventoryChart,
  PeriodicInventoryChartLegend,
  PeriodicRevenueChart,
  PeriodicRevenueChartLegend,
  CumulativeRevenueChart,
  CumulativeRevenueChartLegend,
  CumulativeInventoryChart,
  CumulativeInventoryChartLegend,
  ReportDownloadButton,
  Text
} from '_components';
import {
  cssConstants,
  DATE_FORMATS,
  CHART_HEIGHT,
  GROUP_FILTERS
} from '_constants';
import type { SeasonStatState } from '_state/seasonStat/reducer';
import { selectors as seasonSelectors } from '_state/season';
import { getSeasonStatState } from '_state/seasonStat/selectors';
import typeof SeasonStatActions from '_state/seasonStat/actions';
import type { EDSeason } from '_models';

const TabLink = styled.span`
  color: ${(props) =>
    props.isActive
      ? cssConstants.PRIMARY_LIGHT_BLUE
      : cssConstants.PRIMARY_LIGHT_BLACK};
  font-size: 14px;

  &:hover {
    color: ${cssConstants.PRIMARY_LIGHT_BLUE};
    cursor: pointer;
  }

  & + ${() => TabLink} {
    margin-left: 28px;
  }
`;

const DateFilterOptions = styled.div`
  display: flex;
  align-items: center;

  > * + * {
    margin-left: 6px;
  }
`;

type Props = {
  seasonStatState: SeasonStatState,
  selectedSeason: EDSeason,
  seasonStatActions: SeasonStatActions
};

const NoData = ({ type }: { type: 'Revenue' | 'Inventory' }) => (
  <Flex justify="center" align="center" height={`${CHART_HEIGHT}px`}>
    No {type} Data to Display
  </Flex>
);

const ChartLegend = ({
  selectedGroupFilter,
  selectedTab
}: {
  selectedGroupFilter: number,
  selectedTab: number
}) => {
  if (selectedGroupFilter === 0 && selectedTab === 0)
    return <PeriodicRevenueChartLegend />;
  else if (selectedGroupFilter === 0 && selectedTab === 1)
    return <PeriodicInventoryChartLegend />;
  else if (selectedGroupFilter === 1 && selectedTab === 0)
    return <CumulativeRevenueChartLegend />;
  else if (selectedGroupFilter === 1 && selectedTab === 1)
    return <CumulativeInventoryChartLegend />;
  else return null;
};

export class SeasonRevenuePanel extends React.Component<Props> {
  componentDidMount() {
    const { selectedSeason } = this.props;
    this.props.seasonStatActions.fetch({ seasonId: selectedSeason.id });
  }

  componentWillUnmount() {
    this.props.seasonStatActions.clear();
  }

  componentDidUpdate(prevProps: Props) {
    const { selectedSeason } = this.props;
    if (selectedSeason.id !== prevProps.selectedSeason.id) {
      this.props.seasonStatActions.clear();
      this.props.seasonStatActions.fetch({ seasonId: selectedSeason.id });
    }
  }

  handleDownloadClick = () => {
    const {
      selectedSeason,
      seasonStatState: {
        dateRange: { from, to }
      },
      seasonStatActions: { downloadSeasonReport }
    } = this.props;
    downloadSeasonReport({
      id: selectedSeason.id,
      type: 'season',
      start: from ? from.toISOString() : null,
      end: to ? to.toISOString() : null
    });
  };

  render() {
    const {
      seasonStatState: {
        loading,
        downloading,
        groupFilters,
        dateRange: { from, to },
        dateLimits,
        selectedGroupFilter,
        seasonStats
      },
      seasonStatActions: { setDateRange, setGroupFilter },
      selectedSeason: { startTimestamp, endTimestamp }
    } = this.props;

    const tooltipDateFormat =
      from && to && differenceInCalendarDays(to, from) <= 7
        ? DATE_FORMATS.datetime
        : DATE_FORMATS.day;

    return (
      <Tabbable>
        {(selectedTab, onTabChange) => (
          <Panel>
            <PanelHeader>
              <Flex align="center" justify="space-between" height="100%">
                <FlexItem flex={1}>
                  <TabLink
                    isActive={selectedTab === 0}
                    onClick={() => onTabChange(0)}
                  >
                    Revenue
                  </TabLink>
                  <TabLink
                    isActive={selectedTab === 1}
                    onClick={() => onTabChange(1)}
                  >
                    Inventory
                  </TabLink>
                </FlexItem>
                <FlexItem flex="0" margin="0 0 0 auto">
                  <ChipButtonGroup
                    onChange={setGroupFilter}
                    value={selectedGroupFilter}
                  >
                    {groupFilters.map((group, idx) => (
                      <ChipButton key={idx} value={group.value}>
                        {group.label}
                      </ChipButton>
                    ))}
                  </ChipButtonGroup>
                </FlexItem>
              </Flex>
            </PanelHeader>
            {loading && (
              <PanelContent>
                <Spacing padding="200px 0">
                  <Loader />
                </Spacing>
              </PanelContent>
            )}

            {!loading && (
              <PanelContent>
                <Text size={11} marginBottom="4px">
                  Filter By
                </Text>
                <Flex align="center" justify="space-between">
                  <DateFilterOptions>
                    <DateRangeFilter
                      dateRange={{ from, to }}
                      allTimeDateRange={{
                        from: startTimestamp,
                        to: endTimestamp
                      }}
                      setDateRange={setDateRange}
                      disabledDays={{
                        before: dateLimits.from,
                        after: dateLimits.to
                      }}
                    />
                    <ReportDownloadButton
                      onClick={this.handleDownloadClick}
                      downloading={downloading}
                    />
                  </DateFilterOptions>
                  <FlexItem flex={0}>
                    <Flex justify="flex-end">
                      <ChartLegend
                        selectedGroupFilter={selectedGroupFilter}
                        selectedTab={selectedTab}
                      />
                    </Flex>
                  </FlexItem>
                </Flex>
                <Spacing height="18px" />
                {selectedTab === 0 && (
                  <Fragment>
                    {selectedGroupFilter === GROUP_FILTERS.periodic && (
                      <PeriodicRevenueChart
                        data={seasonStats}
                        height={CHART_HEIGHT}
                        dateFormat={tooltipDateFormat}
                        renderNoData={() => <NoData type="Revenue" />}
                      />
                    )}

                    {selectedGroupFilter === GROUP_FILTERS.cumulative && (
                      <CumulativeRevenueChart
                        data={seasonStats}
                        height={CHART_HEIGHT}
                        dateFormat={tooltipDateFormat}
                        renderNoData={() => <NoData type="Revenue" />}
                      />
                    )}
                  </Fragment>
                )}
                {selectedTab === 1 && (
                  <Fragment>
                    {selectedGroupFilter === GROUP_FILTERS.periodic && (
                      <PeriodicInventoryChart
                        data={seasonStats}
                        height={CHART_HEIGHT}
                        dateFormat={tooltipDateFormat}
                        renderNoData={() => <NoData type="Inventory" />}
                      />
                    )}

                    {selectedGroupFilter === GROUP_FILTERS.cumulative && (
                      <CumulativeInventoryChart
                        data={seasonStats}
                        height={CHART_HEIGHT}
                        dateFormat={tooltipDateFormat}
                        renderNoData={() => <NoData type="Inventory" />}
                      />
                    )}
                  </Fragment>
                )}
              </PanelContent>
            )}
            <PanelContent>
              <RevenueBreakdown
                type={selectedTab === 0 ? 'revenue' : 'inventory'}
              />
            </PanelContent>
          </Panel>
        )}
      </Tabbable>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  seasonStatState: getSeasonStatState,
  selectedSeason: seasonSelectors.selectActiveSeason
});

function mapActionCreators(dispatch) {
  return {
    seasonStatActions: bindActionCreators(seasonStatActions, dispatch)
  };
}

export default connect(
  mapStateToProps,
  mapActionCreators
)(SeasonRevenuePanel);
