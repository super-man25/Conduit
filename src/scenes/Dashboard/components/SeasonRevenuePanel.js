import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import {
  P1,
  Spacing,
  HorizontalBars,
  Panel,
  PanelHeader,
  PanelContent,
  Flex,
  FlexItem,
  LegendItem,
  ChartLegendItem,
  Tabbable,
  Loader,
  SelectDropdown,
  ChipButton,
  ChipButtonGroup
} from '_components';
import { cssConstants } from '_constants';
import { CumulativeRevenueChart } from './CumulativeRevenueChart';
import { isBefore, isSameDay } from 'date-fns';

const selectOptions = [
  { label: 'Seasonal', value: 0 },
  { label: 'Month', value: 1 },
  { label: 'Week', value: 2 }
];

const ChartContainer = styled.div`
  background-color: ${cssConstants.PRIMARY_LIGHTEST_GRAY};
  border: 1px solid ${cssConstants.PRIMARY_LIGHT_GRAY};
  position: relative;
  padding: 14px 14px;
`;

const TabLink = styled.span`
  color: ${(props) =>
    props.isActive
      ? cssConstants.PRIMARY_LIGHT_BLUE
      : cssConstants.PRIMARY_LIGHT_BLACK};
  font-size: 14px;

  &:hover {
    color: ${cssConstants.PRIMARY_LIGHT_BLUE};
    cursor: pointer;
  }

  & + ${() => TabLink} {
    margin-left: 28px;
  }
`;

const RevenueBreakdown = () => (
  <React.Fragment>
    <P1>Revenue Breakdown</P1>
    <Spacing height="12px" />
    <HorizontalBars
      data={[12, 20, 30, 40]}
      colors={['#144670', '#4B98CF', '#4B98CF', '#C4D2E1']}
    />
    <Spacing height="12px" />
    <div>
      <LegendItem label="Single Game" value="$9504.13" color="#144670" />

      <LegendItem label="Group Tickets" value="$9504.13" color="#4B98CF" />

      <LegendItem label="Single Suites" value="$9504.13" color="#2670AE" />

      <LegendItem label="Season Tickets" value="$9504.13" color="#C4D2E1" />
    </div>
  </React.Fragment>
);

export class SeasonRevenuePanel extends React.Component {
  // Temporarily storing these filters in state while still using mock data
  state = {
    groupingType: 'Periodic',
    dateRangeType: selectOptions[0],
    selectedDay: null
  };

  componentDidMount() {
    this.props.eventStatsActions.fetch();
  }

  onGroupingChange = (newType) => {
    this.setState({
      groupingType: newType
    });
  };

  onDateRangeTypeChanged = (dateOption) => {
    this.setState({
      dateRangeType: dateOption
    });
  };

  onDateChanged = (date) => {
    this.setState({
      selectedDay: date
    });
  };

  formatEventStats() {
    const { model } = this.props.eventStatsState;
    const { selectedDay } = this.state;

    const actual = [];
    const projected = [];

    for (const d of model) {
      const formattedModel = {
        ...d,
        x: d.date,
        y: d.revenue
      };

      if (selectedDay && !isSameDay(d.date, selectedDay)) {
        continue;
      }

      if (d.isProjected) {
        projected.push(formattedModel);
      } else {
        actual.push(formattedModel);
      }
    }

    return { actual, projected };
  }

  render() {
    const { groupingType, dateRangeType, selectedDay } = this.state;
    const { loading } = this.props.eventStatsState;

    return (
      <Tabbable>
        {(selectedTab, onTabChange) => (
          <Panel>
            <PanelHeader>
              <Flex align="center" justify="space-between">
                <FlexItem flex={1}>
                  <TabLink
                    isActive={selectedTab === 0}
                    onClick={() => onTabChange(0)}
                  >
                    Revenue Over Time
                  </TabLink>
                  <TabLink
                    isActive={selectedTab === 1}
                    onClick={() => onTabChange(1)}
                  >
                    Inventory / Velocity
                  </TabLink>
                </FlexItem>
                <div>
                  <ChipButtonGroup
                    onChange={this.onGroupingChange}
                    value={groupingType}
                  >
                    <ChipButton value="Periodic">Periodic</ChipButton>
                    <ChipButton value="Cumulative">Cumulative</ChipButton>
                  </ChipButtonGroup>
                </div>
              </Flex>
            </PanelHeader>
            {loading && (
              <PanelContent>
                <Spacing padding="200px 0">
                  <Loader />
                </Spacing>
              </PanelContent>
            )}

            {!loading &&
              selectedTab === 0 && (
                <PanelContent>
                  <Flex align="center" justify="flex-start">
                    <SelectDropdown
                      selected={dateRangeType}
                      options={selectOptions}
                      onChange={this.onDateRangeTypeChanged}
                    />
                  </Flex>
                  <Spacing height="18px" />
                  <ChartContainer>
                    <CumulativeRevenueChart
                      height={400}
                      data={this.formatEventStats()}
                    />
                  </ChartContainer>
                  <Spacing height="18px" />
                  <RevenueBreakdown />
                </PanelContent>
              )}

            {!loading &&
              selectedTab === 1 && (
                <PanelContent>
                  <Flex align="flex-end" justify="flex-end">
                    <ChartLegendItem dashed label="Projected Revenue" />
                    <ChartLegendItem label="Real Revenue" />
                  </Flex>
                  <Spacing height="18px" />
                  <ChartContainer>
                    <CumulativeRevenueChart
                      height={400}
                      data={this.formatEventStats()}
                    />
                  </ChartContainer>
                  <Spacing height="18px" />
                  <RevenueBreakdown />
                </PanelContent>
              )}
          </Panel>
        )}
      </Tabbable>
    );
  }
}

SeasonRevenuePanel.propTypes = {
  eventStatsState: PropTypes.shape(),
  eventStatsActions: PropTypes.shape()
};
