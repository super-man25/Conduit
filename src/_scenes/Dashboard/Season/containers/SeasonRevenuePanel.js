import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { SeasonRevenuePanel } from '_scenes/Dashboard/Season/components/SeasonRevenuePanel';
import { actions as seasonStatActions } from '_state/seasonStat';

function mapStateToProps(state) {
  return {
    seasonStatState: state.seasonStat
  };
}

function mapActionCreators(dispatch) {
  return {
    seasonStatActions: bindActionCreators(seasonStatActions, dispatch)
  };
}

export default connect(mapStateToProps, mapActionCreators)(SeasonRevenuePanel);
