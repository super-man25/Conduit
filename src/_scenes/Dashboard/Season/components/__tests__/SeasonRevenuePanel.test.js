import React from 'react';
import { shallow } from 'enzyme';
import { SeasonRevenuePanel } from '../SeasonRevenuePanel';

describe('<SeasonRevenuePanel />', () => {
  const props = {
    seasonStatState: {
      loading: true,
      groupFilters: [],
      dateRange: { from: null, to: null },
      eventDateLimits: { from: null, to: null },
      selectedGroupFilter: null,
      seasonStats: []
    },
    seasonStatActions: {
      fetch: jest.fn(),
      setGroupFilter: jest.fn(),
      setDateRange: jest.fn()
    }
  };

  it('should render correctly', () => {
    const wrapper = shallow(<SeasonRevenuePanel {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call fetch on componentDidMount', () => {
    shallow(<SeasonRevenuePanel {...props} />);
    expect(props.seasonStatActions.fetch).toHaveBeenCalled();
  });
});
