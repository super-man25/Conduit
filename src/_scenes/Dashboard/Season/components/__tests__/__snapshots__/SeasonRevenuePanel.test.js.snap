// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SeasonRevenuePanel /> should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SeasonRevenuePanel
    seasonStatActions={
        Object {
            "fetch": [Function],
            "setDateRange": [Function],
            "setGroupFilter": [Function],
          }
    }
    seasonStatState={
        Object {
            "dateRange": Object {
              "from": null,
              "to": null,
            },
            "eventDateLimits": Object {
              "from": null,
              "to": null,
            },
            "groupFilters": Array [],
            "loading": true,
            "seasonStats": Array [],
            "selectedGroupFilter": null,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
